### DISCUSSION TITLE
GraphQLXXXTypes implementation

### DISCUSSION THREAD

**Discussion by mihail-makarov-idf** [graphql-java/graphql-java#2578] [Q&A] [answered]
Hi!

We used to be able to inherit GraphQLObjectType, GraphQLFieldDefinition, GraphQLArgument and other GraphQLXXXType classes for our cases (additional fields of third-party classes are added there, which are used to form the data query) . But now we can't do that after you've closed the constructors on those classes in v16 .Now we could implement your interfaces, which implemented the above classes, but in many places in your code there is a cast and check for specific types, not interfaces, for example:
```
 private List<GraphQLObjectType> resolvePossibleObjects(GraphQLCompositeType type, GraphQLSchema graphQLSchema) {
        if (type instanceof GraphQLObjectType) {
            return Collections.singletonList((GraphQLObjectType) type);
        } else if (type instanceof GraphQLInterfaceType) {
            return graphQLSchema.getImplementations((GraphQLInterfaceType) type);
        } else if (type instanceof GraphQLUnionType) {
            List types = ((GraphQLUnionType) type).getTypes();
            return new ArrayList<GraphQLObjectType>(types);
        } else {
            return Assert.assertShouldNeverHappen();
        }

    } 
```

 What can you suggest doing now? Thank you!

**Comment by bbakerman** [comment, marked as answer]
We came to the realisation that we DIDNT want specialisation of these classes.

We do lots of work under the covers to handle "and instance of a graphql object type called "foo" - we want a singleton implementation of them in the schema (literally one object) and we want the class closed.

We also wanted our builders to be the way you create schema elements for correctness reasons.

We could have made is `public final class` say but we didnt.

But in short - we DONT WANT schema element specilisations.

I will ask the reverse question  - what is it that you want to do that means you need your own `SpecialGraphqlObjectType`class

**Reply by yury-vashchyla** [reply]
@bbakerman this is a good case, but it will be executed only after type casts are removed from the code and replaced completely with interfaces. Or it is worth abandoning interfaces and leaving only final classes so as not to mislead developers

**Reply by mihail-makarov-idf** [reply]
@bbakerman We use JOOQ framework to work with the database and have built a graphql-java wrapper that allows us to create graphql objects and fields based on classes generated by JOOQ . In these entities we have added table knowledge, relations (one to many, many to one...) , join conditions. And this is all used in the data fetcher to build a sql query to the database based on the requested fields.

**Reply by bbakerman** [reply]
> @bbakerman We use JOOQ framework to work with the database and have built a graphql-java wrapper that allows us to create graphql objects and fields based on classes generated by JOOQ . In these entities we have added table knowledge, relations (one to many, many to one...) , join conditions. And this is all used in the data fetcher to build a sql query to the database based on the requested fields.

So I think I understand your challenge BUT placing that extra associative meta-data into the actual `GraphQLXXXTypes` objects is not what we plan to allow

So rather than "extending a class" to have the extra data, I think you need to use composition to allow this.

in psudeo code your DataFecther might be

```
DataFetcher df = env -> {

  JooQSystem jooqService= env.getGraphQlContext().get("JOOQCode")
  GraphQLFieldDefinition fDef =  env.getFieldDefinition();
  
  JooqMetaData metaData = jooqService.getMetaDataForField(fDef, env.getParentType(),....)
  
  return  someDataService.getData(env.getSource(), metaData);
  
  }
```

(I used `graphql.GraphQLContext` here to hold this service object - you might have a IoC mechnism to do this better)

The above has look up service code that takes the `GraphQLFieldDefinition` and `GraphQLOutputType` of the parent say as input to look up the associated metadata like tables and so on

So this is "has a " composition versus "is a" inheritance pattern.

**Reply by mihail-makarov-idf** [reply]
> > @bbakerman We use JOOQ framework to work with the database and have built a graphql-java wrapper that allows us to create graphql objects and fields based on classes generated by JOOQ . In these entities we have added table knowledge, relations (one to many, many to one...) , join conditions. And this is all used in the data fetcher to build a sql query to the database based on the requested fields.
> 
> So I think I understand your challenge BUT placing that extra associative meta-data into the actual `GraphQLXXXTypes` objects is not what we plan to allow
> 
> So rather than "extending a class" to have the extra data, I think you need to use composition to allow this.
> 
> in psudeo code your DataFecther might be
> 
> ```
> DataFetcher df = env -> {
> 
>   JooQSystem jooqService= env.getGraphQlContext().get("JOOQCode")
>   GraphQLFieldDefinition fDef =  env.getFieldDefinition();
>   
>   JooqMetaData metaData = jooqService.getMetaDataForField(fDef, env.getParentType(),....)
>   
>   return  someDataService.getData(env.getSource(), metaData);
>   
>   }
> ```
> 
> (I used `graphql.GraphQLContext` here to hold this service object - you might have a IoC mechnism to do this better)
> 
> The above has look up service code that takes the `GraphQLFieldDefinition` and `GraphQLOutputType` of the parent say as input to look up the associated metadata like tables and so on
> 
> So this is "has a " composition versus "is a" inheritance pattern.

Interesting way to look at it. We'll try to redo it that way. Thanks!
